experimental/packages/opentelemetry-instrumentation/src/platform/node/requireInTheMiddleSingleton.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as RequireInTheMiddle from 'require-in-the-middle';
+import * as path from 'path';
+
+export type Hooked = {
+  moduleName: string
+  onRequire: RequireInTheMiddle.OnRequireFn
+};
+
+/**
+ * Whether Mocha is running in this process
+ * Inspired by https://github.com/AndreasPizsa/detect-mocha
+ *
+ * @type {boolean}
+ */
+const isMocha = ['afterEach','after','beforeEach','before','describe','it'].every(fn => {
+  // @ts-expect-error TS7053: Element implicitly has an 'any' type
+  return typeof global[fn] === 'function';
+});
+
+/**
+ * Singleton class for `require-in-the-middle`
+ * Allows instrumentation plugins to patch modules with only a single `require` patch
+ * WARNING: Because this class will create its own `require-in-the-middle` (RITM) instance,
+ * we should minimize the number of new instances of this class.
+ * Multiple instances of `@opentelemetry/instrumentation` (e.g. multiple versions) in a single process
+ * will result in multiple instances of RITM, which will have an impact
+ * on the performance of instrumentation hooks being applied.
+ */
+export class RequireInTheMiddleSingleton {
+  private _modulesToHook: Hooked[] = [];
+  private static _instance?: RequireInTheMiddleSingleton;
+
+  constructor() {
+    this._initialize();
+  }
+
+  private _initialize() {
+    RequireInTheMiddle(
+      // Intercept all `require` calls; we will filter the matching ones below
+      null,
+      { internals: true },
+      (exports, name, basedir) => {
+        // For internal files on Windows, `name` will use backslash as the path separator
+        const normalizedModuleName = normalizePathSeparators(name);
+        const matches = this._modulesToHook.filter(({ moduleName: hookedModuleName }) => {
```

Comment:
Couldn't we precompute this?
From the instrumentation side we essentially know a list of files and functions to patch them:

```js
moduleName: '@nestjs/core'
patch: () => { ... }
version: 'x.y.z'
files: [
  { name: '@nestjs/core/nest-factory.js', patch: () => { ... } }
  { name: '@nestjs/core/router/router-execution-context.js', patch: () => { ... } }
]
```

This list could be flattened to a list of what in function above is `name`, version and a patch function. A little more work up front but saves few passes on an array. Versions can be cached using basedir as key and then we'd not need any complex logic on the instrumentation side.

Reply:
@Rauno56 We now have the `ModuleNameTrie`, which is a trie structure that allows us to avoid array searches.